client/src/main/java/org/eclipse/hono/client/impl/CachingClientFactory.java

```diff
@@ -170,20 +170,29 @@ private void getOrCreateClient(
             creationLocks.put(key, Boolean.TRUE);
             log.debug("creating new client for [{}]", key);
 
-            clientInstanceSupplier.get().setHandler(creationAttempt -> {
+            try {
+                clientInstanceSupplier.get().setHandler(creationAttempt -> {
+                    creationLocks.remove(key);
+                    creationRequests.remove(connectionFailureHandler);
+                    if (creationAttempt.succeeded()) {
+                        final T newClient = creationAttempt.result();
+                        log.debug("successfully created new client for [{}]", key);
+                        activeClients.put(key, newClient);
+                        result.handle(Future.succeededFuture(newClient));
+                    } else {
+                        log.debug("failed to create new client for [{}]", key, creationAttempt.cause());
+                        activeClients.remove(key);
+                        result.handle(Future.failedFuture(creationAttempt.cause()));
+                    }
+                });
+            } catch (final Exception ex) {
```

Comment:
I assume this is supposed to catch an (unexpected runtime) exception thrown by the `clientInstanceSupplier.get()` method invocation, right?

Reply:
Yes, like the one in the description of #1286. 
So this is mainly about preventing the followup `already creating client for key` errors and making the original error more obvious.
See also https://github.com/eclipse/hono/issues/1286#issuecomment-503895722.